### DISCUSSION TITLE
To smurf or not to smurf? That is the question

### DISCUSSION THREAD

**Discussion by KATT** [trpc/trpc#2131] [RFC / Ideas]
> [![brainy-smurf](https://user-images.githubusercontent.com/459267/177989735-97aa4df8-b043-40c6-b92f-ecf2ffaf7b98.png)](https://blog.codinghorror.com/new-programming-jargon/#21)
> When almost every class has the same prefix. IE, when a user clicks on the button, a `SmurfAccountView` passes a `SmurfAccountDTO` to the `SmurfAccountController`. The `SmurfID` is used to fetch a `SmurfOrderHistory` which is passed to the `SmurfHistoryMatch` before forwarding to either `SmurfHistoryReviewView` or `SmurfHistoryReportingView`. If a `SmurfErrorEvent` occurs it is logged by `SmurfErrorLogger` to `${app}/smurf/log/smurf/smurflog.log`
>
> — https://blog.codinghorror.com/new-programming-jargon/#21

@dodas brought up a good question today which lead me to think about our exposed interfaces, should those all contain `TRPC` in the naming somehow?

I'm personally a bit annoyed whenever I use a "generic" name like `createClient` or `createContext` within VSCode, I often don't get the suggestion I want. 

Should tRPC be exposing smurfed interfaces or non-smurfed functions? It would be good to get consistency in this for V10.

### Examples:

- `initTRPC` vs `init`
- `setupNext` vs `setupTRPCNext`
- `setupReact` vs `setupTRPCReact`

**Comment by ixahmedxi** [comment]
smurf ftw

**Comment by sachinraja** [comment]
Smurf all or smurf none (as you noted). I'm fine with smurfing to avoid name collisions and provide better autocomplete.

**Reply by KATT** [reply]
You're _fine_ as in you _prefer_ smurfing?

**Reply by sachinraja** [reply]
To be honest I'm going to backtrack a bit here. I think it's fine to be inconsistent. I would prefer not smurfing as much as possible. However, I think it does make sense where names are more likely to collide like `initTRPC`, `createContext`, `Error`/`TRPCError`. I don't see `setupNext` or `setupReact` being common names so I would keep those as is.

**Reply by KATT** [reply]
I think a reasonable default would be that exposed functions are smurfed. Unsure if `interface`s/`type`s would be overkill?

**Reply by sachinraja** [reply]
:shrug: seems like generally people prefer everything being smurfed

**Reply by mmkal** [reply]
My/my team's I guess minority opinion: I prefer unsmurfed, and use `smurf.` where appropriate to make the source clear and to make autocomplete work:

```ts
import * as trpc from '@trpc/server'

export const foo = trpc.init(...)
export const bar = trpc.router(...)
```

It allows the _option_ of clean, unsmurfed usage when the module name makes the context clear, for example, and smurfed usage otherwise. I find it also helps make clear which functions are portable/reusable/domain-specific which encourages good refactors.

IMO consumer code is better that way too. I much prefer:

```ts
import React from 'react'

export default () => {
  React.useEffect(...)
  return ...
}
```

to

```ts
import { useEffect, useState, useContext, useReducer } from 'react'

export default () => {
  useEffect(...)
  return ...
}
```

Anyway it's not a super strongly held opinion, I'm not going to boycott trpc over having to write `trpc` in a few places.

**Reply by mmkal** [reply]
An example off the top of my head. The `TRPCError` class actually isn't all that coupled to trpc. It's just a class that the trpc server package knows how to deal with. If it _wasn't_ smurfed, and trpc became widely used enough for this to be worthwhile, it could be refactored into a separate package that could be used outside a trpc context. (Tbh, that could still be done, but the "trpc" branding would either have to be removed or carried around with it - which there could be marketing advantages to, admittedly)

**Reply by sachinraja** [reply]
We can't recommend the wildcard imports (and it's not a good idea) because they aren't the best for tree shaking.

**Reply by mmkal** [reply]
I disagree that we can't - I'd hope in 2022 that problem is behind almost everyone. Every major bundler can handle them just fine. Anyone using a super old or super crappy bundler should be considered an exception, and they can always Smurf manually:

```ts
import { init as initTRPC } from '@trpc/server'
```

(or in the case of `@trpc/server`, they can just not worry about it because bundle size doesn't really matter on the server)

**Reply by sachinraja** [reply]
Hmm..you're right. I just tested it with Webpack, esbuild, and Rollup and it correctly did tree shaking. I'm in favor of your suggestion then.

**Reply by KATT** [reply]
I don't like wildcard imports simply because that's not what vscode suggests when you lazily just write code & I don't want people to have to write imports

**Reply by KATT** [reply]
Also, in the new version you'll really mainly do 1 import on the front-end and 1 on the backend which is calling a factory fn that you can declare as whatever you want

**Reply by dodas** [reply]
could we also export wildcard as `trpc`?
e.g.
```tsx
// external/index.ts
export { init } from "./init";
export { router } from "./router";
```

```tsx
// index.ts
export * from "./external";
export * as trpc from "./external";
```

this is basically what [zod does](https://github.com/colinhacks/zod/blob/master/src/index.ts)

**Reply by mmkal** [reply]
I really like that. And to @KATT's point about IDEs/intellisense, I love how typing `z.` is enough for vscode to automatically add the import, and I think this is the reason.

**Reply by sachinraja** [reply]
Not to be that guy again but esbuild failed to tree shake the above (Rollup and Webpack succeeded). Generally I'd rather not do things that reduce the chance of tree shaking happening.

**Reply by mmkal** [reply]
If we follow zod's example, we could `export { trpc }` and people who are using esbuild _and_ don't want to file an issue on esbuild asking why they fail when others succeed, can use wildcard exports. The switch is just `import { trpc } from '@trpc/whatever'` to `import * as trpc from '@trpc/whatever'` on the user side. The cost on this repo's side is a one liner export and (maybe) a note about it in the docs.

**Comment by ChiChuRita** [comment]
I think smurfing is appropriate, especially since it's only 4 letters more. Better autocompletion, no name collisions... as long as we don´t name functions like TypedRemoteProcedureCallContextCreator.

**Reply by samal-rasmussen** [reply]
No smurf please - I like my stuff clean. The namespace is already clear from the import, and if you happen to have a name collision somewhere you can always just add a smurf with `as` in the import.

**Comment by CoenWarmer** [comment]
All good for smurfing.

**Comment by aradalvand** [comment]
Gonna play the devil's advocate: Don't smurf.
Especially because acronyms like tRPC look **extra awkward** in `camelCase`.

And secondarily because it's just tedious to write; brevity ftw.
The context is derived from the import, no need to cram unnecessary words in there.

**Reply by ChiChuRita** [reply]
Thats why we write TRPC not tRPC. (in the smurf context)

**Reply by capaj** [reply]
tedious to write? When using typescript autocomplete is actually more helpful with smurfing. Without it, you might get dozens of unrelated `init` functions etc.

**Comment by iway1** [comment]
Would this mean `TRPCClientError` being renamed `ClientError`? Not sure I like that. 

Smurfing improves readability of the code even though it's uglier. I think it's a mistake to optimize for aesthetics over readability.

Plus, when you have a name space you get some really cool auto complete/auto import functionality. IE you can type "TRPC" and just arrow key through all its exported members. Not sure that's particularly useful here, but just an example of why it might be advantageous.

**Reply by samal-rasmussen** [reply]
I think we must be defining readability very differently because I am much more able to read `ClientError` than `TRPCClientError`.

**Comment by capaj** [comment]
Smurf all the way, everyhwhere. <img src="https://media2.giphy.com/media/5yLgocnCRCdOaBGzQZ2/giphy.gif"/>
